From a1beb0edab8ebbfc5ccafa78059402e350d3cc3f Mon Sep 17 00:00:00 2001
From: Veaceslav Falico <vfalico@redhat.com>
Date: Thu, 21 Aug 2014 11:28:32 -0400
Subject: [ethernet] netxen: Correct off-by-one errors in bounds checks

Message-id: <1408620515-703-8-git-send-email-vfalico@redhat.com>
Patchwork-id: 88514
O-Subject: [PATCH RHEL7 07/10] netxen: Correct off-by-one errors in bounds checks
Bugzilla: 1089356
RH-Acked-by: Tony Camuso <tcamuso@redhat.com>
RH-Acked-by: Stefan Assmann <sassmann@redhat.com>

RHBZ#1089356

Testing:
Locally and by QLogic (which supplied the patch)

Description:
Backport of the following commit:

commit 4710b2ba873692194c636811ceda398f95e02db2
Author: David Gibson <david@gibson.dropbear.id.au>
Date:   Fri Dec 20 15:10:44 2013 +1100

    netxen: Correct off-by-one errors in bounds checks

    netxen_process_lro() contains two bounds checks.  One for the ring number
    against the number of rings, and one for the Rx buffer ID against the
    array of receive buffers.

    Both of these have off-by-one errors, using > instead of >=. The correct
    versions are used in netxen_process_rcv(), they're just wrong in
    netxen_process_lro().

    Signed-off-by: David Gibson <david@gibson.dropbear.id.au>
    Signed-off-by: David S. Miller <davem@davemloft.net>

Signed-off-by: Jarod Wilson <jarod@redhat.com>

diff --git a/drivers/net/ethernet/qlogic/netxen/netxen_nic_init.c b/drivers/net/ethernet/qlogic/netxen/netxen_nic_init.c
index 7692dfd..cc68657 100644
--- a/drivers/net/ethernet/qlogic/netxen/netxen_nic_init.c
+++ b/drivers/net/ethernet/qlogic/netxen/netxen_nic_init.c
@@ -1604,13 +1604,13 @@ netxen_process_lro(struct netxen_adapter *adapter,
  u32 seq_number;
  u8 vhdr_len = 0;
 
- if (unlikely(ring > adapter->max_rds_rings))
+ if (unlikely(ring >= adapter->max_rds_rings))
   return NULL;
 
  rds_ring = &recv_ctx->rds_rings[ring];
 
  index = netxen_get_lro_sts_refhandle(sts_data0);
- if (unlikely(index > rds_ring->num_desc))
+ if (unlikely(index >= rds_ring->num_desc))
   return NULL;
 
  buffer = &rds_ring->rx_buf_arr[index];
-- 
1.7.1